### PR TITLE
[PART 1] Backend changes for Marking Cases as OT

### LINKED ISSUE #13415: Mark Cases as OT
## User or job story
User story: As a Supervisory Senior Counsel (SSC) & Veteran Law Judge (VLJ), I need the ability to designate a case as overtime at the point of assigning to an attorney, so that Attorney productivity/performance can be tracked efficiently. I also need the ability to edit that designation after assignment. 

## Acceptance criteria
- [ ] Please put this work behind the feature toggle: :overtime_revamp
- [ ] This feature should be accessible to the following user groups: VLJs (SSCs are VLJs in Caseflow)
- [ ] Include screenshot(s) in the Github issue if there are front-end changes
- [ ] Update documentation: 
- [ ] When judges or SSCs assign cases to attorneys, they can mark the case as overtime for both AMA and legacy cases
- [ ] Judges or SSCs have the ability to edit the OT designation after assignment
- [ ] Cases marked as OT should display an OT Badge in the Badge Area

### Out of scope
Writing OT designation back to VACOLS from Caseflow

### Current Prototype
Clock icon SVG/HTML can be accessed [here](https://fontawesome.com/icons/clock?style=solid)
[Figma prototype
](https://www.figma.com/proto/qBSLnQhq0H6K7jg6vvW3xW/Mark-AMA-Cases-as-OT?node-id=222%3A0&viewport=254%2C256%2C0.13263243436813354&scaling=min-zoom) ping @geronimoramos if any questions 

### Previous Design mock ups
<details>
<summary>Click to view</summary>
**To mark a case overtime:**
_Senior counsel would click on a blue link in Case Title area of Case Details, reading "Mark this case as overtime":_

image

_Senior counsel can see a confirmation that this case is now marked overtime. They can also see a new badge appear at the top right of case title.:_
image

(The senior counsel will now see another link in Case Title - this time to remove overtime status (in case they made a mistake or have to change the status for other reasons)

_Senior counsel can remove overtime status by clicking on "Remove overtime status for this case"
overtime-marking-remove_
image

they'll be prompted with a pop-up modal message asking them "Are you sure you want to remove overtime status?" ----> if they click "yes" they will see a confirmation reading "Vera Marshall's case is no longer marked as overtime. The "OT" status badge has been removed.
-they will also see that the overtime badge has disappeared, and the blue link now reads "Mark this case as overtime"
</details>

### Technical notes
We are currently only tracking when a case was marked as being worked on OT at the time of the attorney. Any changes to OT-related data will need to be captured.

Overtime data is sent to the following Tableau reporting:
The Attorney production and DAS ratings and comments workbooks

Engineering should inquire with the team that will actually do the work to update Tableau reports what format the OT data should be formatted in.

### Resources/other links
Product Brief: https://docs.google.com/document/d/1OayEvH7KAAojD9XVHkfUHyLdyECC-PkeIsrEx9krqDA/edit#

### PR DESCRIPTION
**Pull request by ajspotts**
# THIS PR

Resolves #13415

PART 1 of the stack
PART 2 #14333 
PART 3 #14332

### Description
Finishes backend work from #14155 by ensuring we properly handle all legacy appeals

### Acceptance Criteria
- [ ] Tests pass

# FULL STACK CHANGES
<details>
<summary></summary>

### Description
Frontend work for #13415

Features added:  
- modal to mark/remove overtime status from case 
- button on Case Details page 

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Work is behind the feature toggle: :overtime_revamp
- [ ] The feature should be accessible to the following user groups: VLJs (SSCs are VLJs in Caseflow)
- [ ] When judges or SSCs assign cases to attorneys, they can mark the case as overtime for both AMA and legacy cases
- [ ] Judges or SSCs have the ability to edit the OT designation after assignment

### Testing Plan

1. enable `overtime_revamp` feature toggle 
```ruby
FeatureToggle.enable!(:overtime_revamp)
```
2. Go to http://localhost:3000/queue as BVAAABSHIRE
3. Select a legacy appeal
4. Ensure you can see the option to Mark this case as overtime
5. Select the option and confirm in the modal the case is to be marked as overtime
6. When the page reloads, ensure that you see the Success banner at the top of the page and you can now see the option to Remove this case from overtime
7. Select the updated option and confirm in the modal the overtime status is to be removed from the case 
8. When the page reloads, ensure that you see the Success banner at the top of the page and you can now see the original option to Mark this case as overtime
9. Repeat steps for an ama appeal
10. Assign one legacy and one ama case each to an attorney on BVAAABSHIRE's team
11.  Ensure that you can successfully repeat steps 4 - 9
12.  Login as whichever attorney you assigned the legacy and ama cases to
13.  Ensure you do not have the option to update the overtime status of the respective cases

### Before

Screen Shot 2020-05-14 at 10 19 32 AM

### After

ezgif com-video-to-gif

### Before

Screen Shot 2020-05-14 at 10 25 32 AM

### After

ezgif com-video-to-gif

### Before

Screen Shot 2020-05-14 at 2 13 17 PM

### After

ezgif com-video-to-gif

### Before

Screen Shot 2020-05-14 at 2 13 36 PM

### After

ezgif com-video-to-gif
</details>

### CODE REVIEW

File: app/models/vacols/case_assignment.rb [hunk 1]
```diff
@@ -26,7 +26,8 @@ def assigned_by
     OpenStruct.new(
       first_name: assigned_by_first_name,
       last_name: assigned_by_last_name,
-      pg_id: assigned_by_user_id
+      pg_id: assigned_by_user_id,
+      css_id: assigned_by_css_id
```
Comment (ajspotts) [inline app/models/vacols/case_assignment.rb:30]: return the `css_id` on the `assigned_by` attribute of legacy tasks

File: app/controllers/work_modes_controller.rb [hunk 2]
```diff
@@ -26,10 +26,23 @@ def appeal
   def validate_modification_access_to_overtime
     fail(Caseflow::Error::ActionForbiddenError) unless FeatureToggle.enabled?(:overtime_revamp)
 
-    unless current_user.judge? && current_user.appeal_has_task_assigned_to_user?(appeal)
+    unless current_user.judge? && current_user_is_assigned_to_appeal?
```
Comment (ajspotts) [inline app/controllers/work_modes_controller.rb:29]: Yoom: Handle case where legacy appeal has an AttorneyLegacyTask assigned by a judge.

File: app/models/user.rb [hunk 3]
```diff
@@ -153,7 +153,7 @@ def fail_if_no_access_to_legacy_task!(vacols_id)
   end
 
   def appeal_has_task_assigned_to_user?(appeal)
-    if appeal.class.name == "LegacyAppeal"
+    if appeal.is_a?(LegacyAppeal)
```
Comment (ajspotts) [inline app/models/user.rb:156]: Yoom: Faster than string comparison.

File: spec/controllers/work_modes_controller_spec.rb [hunk 4]
```diff
@@ -107,22 +107,47 @@
...
     end
 
     context "for legacy appeal" do
-      let(:appeal) { create(:legacy_appeal, vacols_case: create(:case, :assigned, user: judge)) }
+      let(:vacols_case) { create(:case, :assigned, user: judge, as_judge_assign_task: true) }
+      let(:appeal) { create(:legacy_appeal, vacols_case: vacols_case) }
       let(:appeal_id) { appeal.vacols_id }
 
-      let(:root_task) { create(:root_task, appeal: appeal) }
-      let!(:judge_assign_task) { JudgeAssignTask.create!(appeal: appeal, parent: root_task, assigned_to: judge) }
```
Comment (ajspotts) [inline spec/controllers/work_modes_controller_spec.rb:123]: Yoom: Should not have created these AMA tasks for a legacy appeal! Need to add checks in FactoryBot.

File: spec/factories/vacols/case.rb [hunk 5]
```diff
@@ -40,14 +41,21 @@
                    evaluator.work_product
                  end
 
+        # If user=judge and dereceive=nil, then reassigned_to_judge_date=nil, resulting in a JudgeLegacyAssignTask.
+        # If user=judge and dereceive!=nil, then this results in a JudgeLegacyDecisionReviewTask.
+        # Otherwise AttorneyLegacyTask will result.
+        dereceive = if evaluator.user&.vacols_roles&.include?("judge")
+                      evaluator.as_judge_assign_task ? nil : Time.zone.today
+                    end
+
```
Comment (ajspotts) [inline spec/factories/vacols/case.rb:50]: Yoom: I could not find examples to induce a JudgeLegacyAssignTask to be created for a legacy appeal, so I did it this way.

File: spec/controllers/work_modes_controller_spec.rb [hunk 6]
```diff
@@ -107,22 +107,47 @@
...
 
       it_behaves_like "an appeal that can be worked overtime"
+
+      context "when appeal is assigned to attorney" do
+        before { judge_assign_task.completed! }
+
+        let(:judge_review_task) { create(:ama_judge_decision_review_task, parent: root_task, assigned_to: judge) }
+        let!(:attorney_task) { create(:ama_attorney_task, assigned_to: attorney, parent: judge_review_task) }
+
+        it_behaves_like "an appeal that can be worked overtime"
```
Comment (hschallhorn) [inline spec/controllers/work_modes_controller_spec.rb:127]: Nice!

File: spec/factories/vacols/case.rb (hunk 5 above)
Comment (hschallhorn) [reply spec/factories/vacols/case.rb:50]: Man, would love to clean up our factories to make a bit more sense, but this is a great start!

File: spec/controllers/work_modes_controller_spec.rb [hunk 7]
```diff
@@ -150,16 +175,11 @@
...
-          create(
-            :ama_attorney_task,
-            parent: judge_review_task,
-            assigned_by: judge,
-            assigned_to: current_user
-          )
+        let(:judge_review_task) { create(:ama_judge_decision_review_task, parent: judge_assign_task.root_task) }
+        before do
+          judge_assign_task.completed!
+          create(:ama_attorney_task, parent: judge_review_task, assigned_by: judge, assigned_to: current_user)
```
Comment (hschallhorn) [inline spec/controllers/work_modes_controller_spec.rb:181]: Looks like we're still in the legacy task section. [As Yoom pointed out](https://github.com/department-of-veterans-affairs/caseflow/pull/14334/files#r428168141), we shouldn't be creating judge/attorney task s here and instead should be using the decass table though the vacols case.

File: spec/controllers/work_modes_controller_spec.rb [hunk 8]
```diff
@@ -107,22 +107,47 @@
...
       it_behaves_like "an appeal that can be worked overtime"
+
+      context "when appeal is assigned to attorney" do
+        let(:appeal) { create(:legacy_appeal, vacols_case: create(:case, :assigned, user: attorney, assigner: judge)) }
+
+        it "should have an appeal with a AttorneyLegacyTask" do
+          expect(LegacyWorkQueue.tasks_by_appeal_id(appeal.vacols_id).map(&:class)).to eq([AttorneyLegacyTask])
+        end
+
+        it_behaves_like "an appeal that can be worked overtime"
```
Comment (hschallhorn) [inline spec/controllers/work_modes_controller_spec.rb:149]: Doubly nice!

File: app/controllers/work_modes_controller.rb [hunk 9]
```diff
@@ -26,10 +26,23 @@ def appeal
...
+  def current_user_is_assigned_to_appeal?
+    # handle case where legacy appeal has an AttorneyLegacyTask assigned by a judge
+    return true if appeal.is_a?(LegacyAppeal) && any_task_assigned_by_current_user
+
+    current_user.appeal_has_task_assigned_to_user?(appeal)
+  end
+
+  def any_task_assigned_by_current_user
+    tasks = LegacyWorkQueue.tasks_by_appeal_id(appeal.vacols_id)
+    tasks.any? { |task| task.assigned_by[:pg_id] == current_user.id }
```
Comment (hschallhorn) [inline app/controllers/work_modes_controller.rb:46]: `VACOLS::Decass.where(defolder: appeal.vacols_id, demdusr: current_user.vacols_uniq_id).exists?` is a bit quicker and to the point than pulling the whole LegacyWorkQueue for an appeal. 😬 I know I was the one who suggested LegacyWorkQueue
```ruby
Benchmark.measure { VACOLS::Decass.where(defolder: "3662856", demdusr: User.third.vacols_uniq_id).exists? }
=> #<Benchmark::Tms:0x00007fbad5a29d50 @cstime=0.0, @cutime=0.0, @label="", @real=0.09956700005568564, @stime=0.005621000000000098, @total=0.029361000000000193, @utime=0.023740000000000094>

Benchmark.measure do
  tasks = LegacyWorkQueue.tasks_by_appeal_id("3662856")
  tasks.any? { |task| task.assigned_by[:pg_id] == 3 }
end
=> #<Benchmark::Tms:0x00007fbae4301258 @cstime=0.0, @cutime=0.0, @label="", @real=1.9085039999336004, @stime=0.0965600000000002, @total=0.7193449999999988, @utime=0.6227849999999986>
```

File: spec/controllers/work_modes_controller_spec.rb (hunk 7 above)
Comment (ajspotts) [reply spec/controllers/work_modes_controller_spec.rb:181]: I think the section for legacy appeals actually only between lines 131 - 151 and the section beginning on 153 `when non-judge user modifies overtime` is pertaining to only ama appeals

File: app/controllers/work_modes_controller.rb (hunk 9 above)
Comment (ajspotts) [reply app/controllers/work_modes_controller.rb:46]: @yoomlam suggested that we could also extract your suggestion to `QueueRespository` where we're already doing [similar calls](https://github.com/department-of-veterans-affairs/caseflow/blob/57e746e91b91c82c9b5f4edac343bde9688b68d9/app/repositories/queue_repository.rb#L150).  If this works for you I'll go about this implementation.

File: spec/controllers/work_modes_controller_spec.rb (hunk 7 above)
Comment (hschallhorn) [reply spec/controllers/work_modes_controller_spec.rb:181]: true true true

File: app/controllers/work_modes_controller.rb (hunk 9 above)
Comment (hschallhorn) [reply app/controllers/work_modes_controller.rb:46]: Great idea! Works for me!

File: app/repositories/queue_repository.rb [hunk 10]
```diff
@@ -168,6 +168,10 @@ def reassign_case_to_attorney!(judge:, attorney:, vacols_id:, created_in_vacols_
       end
     end
 
+    def any_task_assigned_by_user?(user, appeal)
```
Comment (hschallhorn) [inline app/repositories/queue_repository.rb:171]: Crazy nitpick, but could we swap user and appeal here?

Comment (ajspotts) [reply app/repositories/queue_repository.rb]: swapped!

**Pull request merged**